### PR TITLE
fix(dynamics): report counter-based graduation in FusedStage step()

### PR DESCRIPTION
**Pull request by lil-lon**
# ALCHEMI Toolkit Pull Request

## Description

`FusedStage.step()` is documented to return the indices of samples that newly reached `exit_status` during the step, but samples graduated via a sub-stage's `n_steps` counter were never reported: the `pre_converge_status` snapshot was taken after the counter-migration loop, so those samples already had `status == exit_status` in the snapshot and were excluded from the `newly_graduated` comparison. Consumers of the returned indices (e.g. `DistributedPipeline`) silently dropped such samples.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

No issue is reported; the minimal reproducible code is as follows.

```python
import torch
from ase.build import molecule

from nvalchemi.data import AtomicData, Batch
from nvalchemi.dynamics import FIRE, FusedStage
from nvalchemi.models.demo import DemoModel, DemoModelWrapper


def atoms_to_data(atoms) -> AtomicData:
    data = AtomicData.from_atoms(atoms)
    n = data.num_nodes
    data.forces = torch.zeros(n, 3)
    data.energy = torch.zeros(1, 1)
    data.add_node_property("velocities", torch.zeros(n, 3))
    return data


model = DemoModelWrapper(DemoModel())

# One sub-stage with a 2-step budget and NO convergence hook:
# graduation happens exclusively via the n_steps counter.
fire = FIRE(model=model, dt=0.1, n_steps=2)
fused = FusedStage(sub_stages=[(0, fire)], max_batch_size=4)
print(f"exit_status = {fused.exit_status}")

batch = Batch.from_data_list([atoms_to_data(molecule("H2O"))])

for i in range(1, 3):
    batch, exit_converged = fused.step(batch)
    print(
        f"step {i}: status={batch.status.view(-1).tolist()}, "
        f"exit_converged={exit_converged}"
    )
```

## Changes Made

- Move the `pre_converge_status` snapshot above the `n_steps` counter-migration loop in `FusedStage._step_impl`, so both migration paths (counter and convergence hook) are covered by the `newly_graduated` comparison.
- Add a regression test: a single-sub-stage `FusedStage` with `n_steps=2` and no convergence hook must return non `None` `exit_converged` on the step the counter graduates the sample.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

Note on contribution timing: I'm aware this repository is currently in its public-beta phase and not yet accepting direct code contributions.
I opened this PR because the fix is small and self-contained, but I'm happy to close this PR in favor of creating an issue (or hold it) if you'd prefer. Please just let me know which you'd prefer.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/dynamics/base.py [hunk 1]
```diff
@@ -3000,6 +3000,11 @@ def _step_impl(self, batch: Batch) -> tuple[Batch, torch.Tensor | None]:
         self._call_hooks(DynamicsStage.AFTER_STEP, batch)
         self._call_fused_hooks(DynamicsStage.AFTER_STEP, batch)
 
+        # Snapshot before the counter migration so it is reported in exit_converged.
+        pre_converge_status = batch.status.clone()
```
Comment (dallasfoster) [inline nvalchemi/dynamics/base.py]: Can we rename to `pre_migration_status`

**Review by dallasfoster — CHANGES_REQUESTED** [review]
Consider adding a multi-sample batch where some samples graduate via the counter and others via a convergence hook in the *same* step would exercise the combined path in one assertion.

**Comment by lil-lon** [comment]
Thanks for suggesting the combined test! While writing it, I found that hook-based graduation was not reported either. So, I fixed it and modified the change log accordingly: [a31d906](https://github.com/NVIDIA/nvalchemi-toolkit/pull/131/commits/a31d9060e15cc2ef0528410408a0c70a300c4d37)
Also, I modified the test to assert the combined path as you suggested: [20bac50](https://github.com/NVIDIA/nvalchemi-toolkit/pull/131/commits/20bac50f72a7db44b34a9bd28380e193cb68a63b)

File: nvalchemi/dynamics/base.py (hunk 1 above)
Comment (lil-lon) [reply nvalchemi/dynamics/base.py]: Renamed in [a31d906](https://github.com/NVIDIA/nvalchemi-toolkit/pull/131/commits/a31d9060e15cc2ef0528410408a0c70a300c4d37)

**Review by dallasfoster — APPROVED** [review]

**Comment by dallasfoster** [comment]
/ok to test [23339ec](https://github.com/NVIDIA/nvalchemi-toolkit/pull/131/commits/23339ec354495e5a26c847b28ec3902432f2442e)
